msal_extensions/osx.py

```diff
@@ -0,0 +1,315 @@
+# pylint: disable=duplicate-code
+
+"""Implements a macOS specific TokenCache, and provides auxiliary helper types."""
+
+import os
+import ctypes as _ctypes
+
+OS_RESULT = _ctypes.c_int32
+
+
+class KeychainError(OSError):
+    """The parent of all exceptions that may be thrown by the functions implemented inside of the
+    Keychain type."""
+    def __init__(self, exit_status):
+        super(KeychainError, self).__init__()
+        self.exit_status = exit_status
+        # TODO: pylint: disable=fixme
+        #  use SecCopyErrorMessageString to fetch the appropriate message here.
+        self.message = \
+            '{} ' \
+            'see https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/MacErrors.h'\
+                .format(self.exit_status)
+
+
+class KeychainAccessDeniedError(KeychainError):
+    """The exceptions that's raised when a Keychain is set to be queried, but the current process
+    has not been give permission to access it, and a query comes anyway."""
+    EXIT_STATUS = -128
+
+    def __init__(self):
```

Comment:
So the first 70+ lines here essentially implement an exception hierarchy to represent the flat underlying error codes. While such motivation is understandable (so that the downstream developers would only need to catch a descriptive exception, not needing to know and compare a specific errno value), the real benefit seems only meaningful when/if a downstream developer would specifically catch and handle a particular subset of these exceptions but not the others. Otherwise, there is no difference than always raising the same base `KeychainError` which already contains the `exit_status` and, to Martin's credit, a hint pointing to Apple's official single source of truth. Downstream developer's life would be exactly the same when we raise only `KeychainError` and not its children.

If, however, you still insist to do the exception tree here, that's also fine. In such case the suggestion would become more tactical. We should keep the same `__init__(self, exit_status)` here (otherwise we are violating the L of [SOLID principle](https://en.wikipedia.org/wiki/SOLID)). And then in this particular case, that means we can completely remove a duplicated `__init__(...)` implementation here.

Reply:
Oooh, I love this. I would argue that this is not a violation of the 'L' from SOLID. 

Basically, my position is that adding the argument `exit_status` here would be misleading, as the only acceptable value for an AccessDeniedError is `-128`. Further more, I believe Liskov substitution is only talking about the properties of instantiated objects. At least in other languages, I haven't run into examples of needing to have covariant or contravariant constructors! Let me know if Python is doing something unknown to me that changes my thinking. :)

Reply:
Wow alright, well I don't think my argument above is wrong! However, seeing the [syntax sugar](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15/files/d3059c37617d75d1c583615fc65ad95e3036a99e#r292215210) your approach unlocks, I think it warrants having a slightly misleading constructor. :) Especially as you say [here](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15/files/d3059c37617d75d1c583615fc65ad95e3036a99e#r292214519), we'll be the only ones to ever have to actually invoke it. My strong position was weakly held, and I am yielding it!

Reply:
Fixed in 2962bd8d69a24544a0f6f725e7e0e89cb16fbcea

Reply:
Note, I did add `assert` entries. This way we have some certainty that the correct value is being injected, plus people can [invoke python in such a way that these assertions disappear anyway](https://docs.python.org/3/using/cmdline.html#cmdoption-o).

Reply:
Resolving this conversation after [offline discussion](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15#issuecomment-505079762)